dhis2/dhis2-core
# chore: DataValueSet writer API [DHIS2-11943]

### Summary
Introduces a `DataValueSetReader` (for import) and a `DataValueSetWriter` (for export) abstraction.

These contain the import/export relevant code that previously was "bolted on" the `DataValueSet` and `DataValue` using inheritance. This means both  `DataValueSet` and `DataValue` are now just POJOs with no further subclassing. 

The import and export now operate on the new `DateValueEntry` interface. This is implemented by readers and writers to directly expose their data from the source input stream in a uniform representation that the import and export generic code can process it. The reader and writer support and use the try-with-resource construct to make sure the streams are closed.

As a consequence there now is less coping between different representations and a more clear separation of concerns. The POJOs are just data records, import has a reader class per input format, export has a writer class per format.

Reading supports both streaming as well as full mapping.
Like before we currently fully map JSON and PDF to a `DataValueSet` that includes a list of `DataValue`s while XML and CSV use a streaming approach where the `DataValueSet` "header" does not include any `DataValue` - instead these are iterated over directly reading them from the input stream. This also prepares us to switch over JSON to streaming based in the future which I think we can do using https://github.com/dhis2/json-tree.

I also renamed the service methods to `import...` and `export...` as I got confused multiple times by `save...` and `write...`. 

### Automated Testing
There are numerous existing tests that from past experience fail easily even when details are slightly changed.

### Manually Testing
* open Import/Export app, Data value import/export
* export to all combinations of format and compression and check the files contain sensible data
* import the uncompressed files exported using the import and make a dry run, see if the results make sense
